[FLINK-12716][python] Add an interactive shell for Python Table API

HuangXingBo opened the pull request:
## What is the purpose of the change
In this PR, we want to add an interactive shell for the Python Table API. It will have the similar functionality like the Scala Shell.
## Brief change log

  - *Add shell.py in flink-python/pyflink as the starting script in python shell*
  - *Add bin/pyflink-shell.sh as the startting script of python shell*
  - *Add class PythonShellParser to parse python shell options*
  - *correct bin/pyflink-gateway-server.sh* 


## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

dianfu commented on flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh:
```diff
@@ -67,4 +67,16 @@ if [[ -n "$FLINK_TESTING" ]]; then
 fi
 
-exec $JAVA_RUN $JVM_ARGS "${log_setting[@]}" -cp ${FLINK_CLASSPATH}:${TABLE_JAR_PATH}:${PYTHON_JAR_PATH}:${FLINK_TEST_CLASSPATH} ${DRIVER} ${ARGS[@]}
+COUNT=${#ARGS[@]}
```
COUNT -> ARGS_COUNT

dianfu commented on flink-clients/src/main/java/org/apache/flink/client/cli/PythonShellParser.java:
```diff
@@ -0,0 +1,288 @@
+	 * @param args Python shell yarn options.
+	 * @return Yarn options usrd in `flink run`.
+	 */
+	private static List<String> parseYarn(String[] args) {
```
The yarn mode doesn't work: The following exception will be thrown:
JAR file does not exist: yarn-cluster

dianfu commented on flink-dist/src/main/flink-bin/bin/pyflink-gateway-server.sh:
```diff
@@ -67,4 +67,16 @@ if [[ -n "$FLINK_TESTING" ]]; then
+  exec $JAVA_RUN $JVM_ARGS "${log_setting[@]}" -cp ${FLINK_CLASSPATH}:${TABLE_JAR_PATH}:${PYTHON_JAR_PATH}:${FLINK_TEST_CLASSPATH}:${EXTERAL_CLASSPATH} ${DRIVER} ${ARGS[@]}
+else
+  ARGS=("${ARGS[@]:1:$COUNT}")
+  exec "$FLINK_BIN_DIR"/flink run ${ARGS[@]} -c ${DRIVER} -a ${PYTHON_JAR_PATH} -j ${TABLE_JAR_PATH}
```
What about removing the option `-a`?
We can set the `isPython` to true if the entrypoint class is "PythonGatewayServer" in `ProgramOptions`.

HuangXingBo replied:
Yes.It sounds better.

dianfu commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,182 @@
+
+## Usage
+
+The shell supports Table API currently. 
```
only supports

dianfu commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,182 @@
+## Usage
+
+The shell supports Table API currently. 
+The Table Environment are automatically prebound after startup. 
```
Environments

dianfu commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,182 @@
+
+### Remote
+
+To use it with a running cluster start the Python shell with the keyword `remote`
```
comma before start

WeiZhong94 commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,182 @@
+<div class="codetabs" markdown="1">
+<div data-lang="batch" markdown="1">
+{% highlight python %}
+>>>import tempfile
```
How about add a space " " between ">>>" and the python code？

WeiZhong94 commented on flink-dist/src/main/flink-bin/bin/pyflink-shell.sh:
```diff
@@ -0,0 +1,79 @@
+opt=`dirname "$0"`
+opt=`cd "$opt"/../opt; pwd`
+export PYTHONPATH="$opt/python/pyflink.zip:$PYTHONPATH"
+PY4J_ZIP=`echo "$opt/python/py4j-*-src.zip"`
```
The star character in double quote won‘t expand, how about remove the double quotes or only quotes the "$opt"?

WeiZhong94 commented on flink-python/pyflink/shell.py:
```diff
@@ -0,0 +1,100 @@
+    platform.python_build()[0],
+    platform.python_build()[1]))
+
+# TODO: refine the sample
```
How about remove this？

HuangXingBo commented on flink-dist/src/main/flink-bin/bin/pyflink-shell.sh:
```diff
@@ -0,0 +1,79 @@
+opt=`dirname "$0"`
+opt=`cd "$opt"/../opt; pwd`
+export PYTHONPATH="$opt/python/pyflink.zip:$PYTHONPATH"
+PY4J_ZIP=`echo "$opt/python/py4j-*-src.zip"`
```
Yes.At first,I didn't add "".Thank you for finding this error.

HuangXingBo commented on flink-clients/src/main/java/org/apache/flink/client/cli/PythonShellParser.java:
```diff
@@ -0,0 +1,288 @@
+ * A simple command line parser (based on Apache Commons CLI) that extracts command
+ * line options.The parser will be used in python shell.
+ */
+public class PythonShellParser {
```
At beginning, I considered this parser class can replace the similar logic in Flink scala shell.But currently, this class is only used in python shell.Maybe in the future, we can reconstruct the flink scala shell to use this class.

HuangXingBo commented on flink-dist/src/main/flink-bin/bin/pyflink-shell.sh:
```diff
@@ -0,0 +1,79 @@
+
+# Add pyflink & py4j to PYTHONPATH
+opt=`dirname "$0"`
+opt=`cd "$opt"/../opt; pwd`
```
Yes.I don't have to align with the scala-shell script.

HuangXingBo commented on flink-dist/src/main/flink-bin/bin/pyflink-shell.sh:
```diff
@@ -0,0 +1,79 @@
+
+FLINK_CLASSPATH=`constructFlinkClassPath`
+
+PYFLINK_INTERPRETER="${PYFLINK_INTERPRETER:-"python"}"
```
It sounds better

HuangXingBo commented on flink-python/pyflink/java_gateway.py:
```diff
@@ -66,6 +67,9 @@ def launch_gateway():
     command = [os.path.join(FLINK_HOME, script)]
     command += ['-c', 'org.apache.flink.python.client.PythonGatewayServer']
 
+    submit_args = os.environ.get("SUBMIT_ARGS", "local")
```
In the script python-gateway-server.sh

HuangXingBo commented on flink-clients/src/main/java/org/apache/flink/client/cli/PythonShellParser.java:
```diff
@@ -0,0 +1,288 @@
+	private static final Option ADD_JARS_OPTION = new Option("a", "addclasspath", true,
+		"Specifies additional jars to be used in Flink.");
+
+	private static final Option container = new Option("n", "container", true,
```
Yes. The build way of CliOptionsParser looks better.

zjffdu commented on flink-dist/src/main/flink-bin/bin/pyflink-shell.sh:
```diff
@@ -0,0 +1,79 @@
+PY4J_ZIP=`echo "$opt/python/py4j-*-src.zip"`
+export PYTHONPATH="$PY4J_ZIP:$PYTHONPATH"
+
+DRIVER="org.apache.flink.client.cli.PythonShellParser"
```
How about rename `PythonShellParser` to `PythonShellDriver` ?

HuangXingBo replied:
Thank you @zjffdu . PythonShellParser is only responsible for parsing the options.So in my opinion, Calling PythonShellParser might be better.How do you think of this？

zjffdu replied:
I didn't check this class in details, just notice you assign to `DRIVER`, so guess maybe `PythonShellDriver` is better.

HuangXingBo replied:
I will change DRIVER to PARSER to avoid misleading.Thank you.

zjffdu commented:
I tried this PR, but go this error when starting pyflink shell

```
![image](https://user-images.githubusercontent.com/164491/59258076-d5ec5200-8c69-11e9-9e21-a0721c1072bd.png)
```

Here's what I did:

* mvn install -Pfast -Pinclude-hadoop -DskipTests  (Build flink)
* bin/pyflink-shell.sh local

HuangXingBo commented:
@zjffdu This error is caused by the unresolved * in the double quotes mentioned earlier. I fixed this problem when I updated it. You can update the code and try again.

HuangXingBo commented on flink-clients/src/main/java/org/apache/flink/client/cli/PythonShellParser.java:
```diff
@@ -0,0 +1,288 @@
+ * A simple command line parser (based on Apache Commons CLI) that extracts command
+ * line options.The parser will be used in python shell.
+ */
+public class PythonShellParser {
```
In my opinion, Because CliFrontendParser and CliOptionParser are not final class, the PythonShellParser is a common class may be okay.How do you think?

HuangXingBo commented:
Thank you @dianfu @WeiZhong94 I have updated the code according to your comments.

zjffdu commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,190 @@
+>>>     .register_table_sink("batch_sink")
+>>> t.select("a + 1, b, c")\
+>>>     .insert_into("batch_sink")
+>>> bt_env.execute()
```
These 2 examples are too complicated for python shell. IMHO, the examples in python shell should be as simple as possible. Such as following:
```
t = bt_env.from_elements([(1, 'hi', 'hello'), (2, 'hi', 'hello')], ['a', 'b', 'c'])
t.print()
t.select("a + 1, b, c").print()
```

HuangXingBo replied:
Currently.Table doesn't have the print method both Java and Python.The example is currently easiest example.In the future, we will add print method and improve this example.

zjffdu replied:
Then is there any other api like `collect` or `head` ? To be honest, this kind of api is too complicated for python users.

HuangXingBo replied:
In Java, Table is first converted to DataSet or DataSteam. Methods such as print are then called for output. However, Python api has not yet implemented the content of DataSet and DataStream, so there is no way to use it this way. When we have completed the corresponding api, we can operate on the python side just like the Java side. So it can be thought that the current example still needs to be improved in the future, but for now it can only be used in this way.There is no doubt that this kind of use is very troublesome now.

zjffdu replied:
Thanks for the clarification. Let's improve it in another PR.

HuangXingBo replied:
Thanks a lot for your suggestions.

dianfu commented:
LGTM now. +1

sunjincheng121 commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,185 @@
+>>> sink_path = tempfile.gettempdir() + '/streaming.csv'
+>>> if os.path.isfile(sink_path):
+>>>     os.remove(sink_path)
+>>> t = st_env.from_elements([(1, 'hi', 'hello'), (2, 'hi', 'hello')], ['a', 'b', 'c'])
```
I think we can set the `set_parallelism(1)`, then we can check the result easily, What to do you think?

sunjincheng121 commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,185 @@
+>>>     .register_table_sink("stream_sink")
+>>> t.select("a + 1, b, c")\
+>>>     .insert_into("stream_sink")
+>>> st_env.execute()
```
It's better to tell the user how to check the result. currently, I think we can give the user some example to read the sink file, What do you think?

sunjincheng121 commented on flink-python/src/main/java/org/apache/flink/python/client/PythonShellParser.java:
```diff
@@ -0,0 +1,322 @@
+	/**
+	 * Prints the help for the client.
+	 */
+	private static void printHelp() {
```
I think we should have deffrence help message for `local`, `remote` and `yarn`, otherwise we do not need add the `-h` option for `local`, `remote` and `yarn`. What do you think?

sunjincheng121 commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,185 @@
+{% endhighlight %}
+
+
+## Full Reference
```
The doc of the `full reference` should some as the command line output`:
![image](https://user-images.githubusercontent.com/22488084/59479344-42946600-8e8f-11e9-9afc-7c3b5d7415e6.png)

sunjincheng121 commented on docs/ops/python_shell.md:
```diff
@@ -0,0 +1,185 @@
+>>> import os
+>>> sink_path = tempfile.gettempdir() + '/batch.csv'
+>>> if os.path.isfile(sink_path):
+>>>     os.remove(sink_path)
```
I think we should not only remove the file, but alos need delete the folder by `shutil.rmtree(result_path)`, What do you think?

sunjincheng121 commented on docs/ops/python_shell.zh.md:
```diff
@@ -0,0 +1,179 @@
+bin/pyflink-shell.sh local
+{% endhighlight %}
+
+关于如何在一个Cluster集群上运行Python shell，可以参考下面的启动那一节的内容
```
`关于如何在一个Cluster集群上运行Python shell，可以参考下面的启动那一节的内容` -> `关于如何在一个Cluster集群上运行Python shell，可以参考启动章节介绍。`

sunjincheng121 commented on docs/ops/python_shell.zh.md:
```diff
@@ -0,0 +1,179 @@
+
+### Table API
+
+下面的内容是关于如何通过Python Table API来实现一个简单的作业:
```
下面是一个通过Python Shell 运行的简单示例。

sunjincheng121 commented on docs/ops/python_shell.zh.md:
```diff
@@ -0,0 +1,179 @@
+
+## 启动
+
+为了概览Python Shell提供的可选参数，可以使用:
```
查看Python Shell提供的可选参数，可以使用:

sunjincheng121 commented on docs/ops/python_shell.zh.md:
```diff
@@ -0,0 +1,179 @@
+
+### Local
+
+可以指定Python Shell运行在local模式下，只需要执行:
```
Python Shell运行在local模式下，只需要执行:

sunjincheng121 commented on docs/ops/python_shell.zh.md:
```diff
@@ -0,0 +1,179 @@
+
+### Remote
+
+可以指定Python Shell运行在一个指定的JobManager上，通过关键字`remote`和对应的JobManager
```
We can delete `可以指定`.

sunjincheng121 commented on docs/ops/python_shell.zh.md:
```diff
@@ -0,0 +1,179 @@
+
+### Yarn Python Shell cluster
+
+可以指定Python Shell运行在YARN集群之上。YARN的container的数量可以通过参数`-n <arg>`进行
```
`可以指定Python Shell运行在YARN集群之上` -> `Python Shell可以运行在YARN集群之上。`

sunjincheng121 commented on flink-python/pyflink/shell.py:
```diff
@@ -0,0 +1,128 @@
+    * t.select("a + 1, b, c")\\
+    *   .insert_into("batch_sink")
+    *                
+    * bt_env.execute()
```
should add the code to check the result file.

sunjincheng121 commented on flink-python/pyflink/shell.py:
```diff
@@ -0,0 +1,128 @@
+    * import os
+    * sink_path = tempfile.gettempdir() + '/batch.csv'
+    * if os.path.isfile(sink_path):
+    *   os.remove(sink_path)
```
should check the folder, and using `shutil.rmtree()` delete the folder.

sunjincheng121 commented:
Thanks for the update @HuangXingBo 👍 
+1 to merged!
